### PR TITLE
Refactor documentation lexer and parser

### LINKED ISSUE #25279: Refactor documentation parser and Lexer
**Description:**

1. Move commonly used lexer methods to AbstractLexer
2. Remove global variables used in DocumentationParser
3. Give a parser warning when `# Deprecation` is followed by a documentation description token.  
4. Support new identifier definition which is subject to [1] .
```
identifier := UnquotedIdentifier | QuotedIdentifier
UnquotedIdentifier := (IdentifierInitialChar | IdentifierEscape) (IdentifierFollowingChar | IdentifierEscape)*
QuotedIdentifier := ' (IdentifierFollowingChar | IdentifierEscape)+
``` 

**Suggested Labels (optional):**
Component/Parser, Type/Improvement

**Suggested Assignees (optional):**
@lochana-chathura

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
- [x] Move commonly used lexer methods to AbstractLexer. e.g. isIdentifierFollowingChar()
- [x] Remove global variables used for char-sequence-search in DocumentationLexer.
- [x] Add documentation support for the new identifier definition.
- [ ] Give a parser warning when "# Deprecated" is followed by more text. 
( Will not address here as it is still in discussion ballerina-platform/ballerina-spec#461. But improved formatting for such scenario)
- [x] Fix ability of using single quote alone as a quoted identifier.

Fixes #25279

## Approach
N/A

## Samples
Sample with documenting quoted identifier
```bal
# Adds parameter `'foo_ƮέŞŢ` and parameter `'üňĩćőđę_param`
# + 'foo_ƮέŞŢ - one thing to be added
# + 'üňĩćőđę_param - another thing to be added
# + return - the sum of them
function add (int 'foo_ƮέŞŢ, int 'üňĩćőđę_param) returns int {
     return 'foo_ƮέŞŢ + 'üňĩćőđę_param;
}
```
## Remarks

- In documentation lexer, we did a char-sequence-search before tokenizing the content within the backticks.
- Now with the new identifier definition searching it at lexer level seemed to be more expensive.
- Therefore, we perform a token-sequence-search in the parser level instead.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java [hunk 1]
```diff
@@ -266,10 +378,81 @@ private STNode parseBacktickToken() {
...
+        if (isValidBacktickContentSequence(referenceGenre)) {
+            return parseBacktickContent();
+        }
+
+        STNode contentToken = combineAndCreateBacktickContentToken();
+        if (referenceGenre != ReferenceGenre.NO_KEY) {
+            // Log warning for backtick block with a reference type, but content is invalid.
+            contentToken = SyntaxErrors.addDiagnostic(contentToken,
+                    DiagnosticWarningCode.WARNING_INVALID_DOCUMENTATION_IDENTIFIER, ((STToken) contentToken).text());
+        }
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java:397]: Don't we give a warning for invalid backtick content with `NO_KEY`. Also `combineAndCreateBacktickContentToken` can be renamed to `combineAndCreateInvalidBacktickContentToken`

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/AbstractLexer.java [hunk 2]
```diff
@@ -134,4 +134,117 @@ protected STNode getLeadingTrivia() {
...
+     * <code>IdentifierInitialChar := A .. Z | a .. z | _ | UnicodeIdentifierChar</code>
+     *
+     * @param c character to check
+     * @return <code>true</code>, if the character is an identifier start char. <code>false</code> otherwise.
+     */
+    protected static boolean isIdentifierInitialChar(int c) {
+        if ('A' <= c && c <= 'Z') {
+            return true;
+        }
+
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/AbstractLexer.java:192]: We can remove these new lines.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationLexer.java [hunk 3]
```diff
@@ -117,56 +107,97 @@ private String getLexeme() {
...
-        return isIdentifierInitialChar(c) || isDigit(c);
+    private void processIdentifierEnd(boolean initialEscape) {
+        while (!reader.isEOF()) {
+            int k = 1;
+            int nextChar = reader.peek();
+            if (isIdentifierFollowingChar(nextChar)) {
+                reader.advance();
+                continue;
+            }
+
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationLexer.java:143]: Can remove the new line here as well

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java [hunk 4]
```diff
@@ -157,13 +158,124 @@ private STNode parseDocumentationReference() {
...
+                break;
+            case NO_KEY:
+                // Look for x(), m:x(), T.y(), m:T.y() match
+                hasMatch = hasBacktickExpr(lookahead, false);
+                break;
+            default:
+                throw new IllegalStateException("Unsupported backtick reference genre");
+        }
+
+        return hasMatch && peek(lookahead.offset).kind == SyntaxKind.BACKTICK_TOKEN;
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java:207]: If we can return `false` from the calling methods ie `hasQualifiedIdentifier` and `hasBacktickExpr` whenever `peek(lookahead.offset).kind == SyntaxKind.BACKTIC_CONTENT` then we can simplify this by removing the `Lookahead` class

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java (hunk 1 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java:397]: We need to clarify it with the spec. There were several reference like above in repo atm. Create an issue for that. #27958
+1 on changing the name. Will do once clarified.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/AbstractLexer.java (hunk 2 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/AbstractLexer.java:192]: I think leaving a newline after a block whenever appropriate is a good practice. Ref: https://github.com/ballerina-platform/ballerina-lang/pull/26717#discussion_r527414794 also.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java (hunk 4 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/DocumentationParser.java:207]: As we discussed offline problem is, methods like hasIdentifier is called in other places as well. eg. hasQualifiedIdentifer, hasBacktickExpr.

**Review by rdulmina — APPROVED** [review]

**Pull request merged**